### PR TITLE
add flatten_grad spmd rule and modify flatten spmd rule to call spmd rule of reshape

### PR DESCRIPTION
**Pull request by jeff41404**
### PR Category
Auto Parallel

### PR Types
New features

### Description
pcard-84225
- add flatten_grad spmd rule
- modify flatten spmd rule to call spmd rule of reshape, so can support dygraph

### CODE REVIEW

File: test/cpp/auto_parallel/spmd_rule_test.cc [hunk 1]
```diff
@@ -1853,6 +1853,61 @@ TEST(CumSumGradInferSpmd, Ctor) {
...
+  // [b, h/ph, w/pw, c, ph, pw] => [b, h/ph, w/pw, hidden_size]
+  auto spmd1 = FlattenInferSpmd(input1, -3, -1);
+  EXPECT_EQ(spmd1.first.size(), static_cast<size_t>(1));
+  EXPECT_EQ(spmd1.second.size(), static_cast<size_t>(2));
+  check_dim_mapping(spmd1.first[0], {0, -1, -1, -1, -1, -1});
+  check_dim_mapping(spmd1.second[0], {0, -1, -1, -1});
+  check_dim_mapping(spmd1.second[1], {-1, 0, -1, -1, -1, -1, -1});  // x_shape
+
+  // [b, h/ph, w/pw, c, ph, pw]; dp, mp
+  auto input2 = build_input({4, 16, 16, 4, 2, 2}, {-1, 0, -1, 1, -1, -1});
```
Comment (JZ-LIANG) [inline test/cpp/auto_parallel/spmd_rule_test.cc:1884]: 最好再补一个单测场景：  切分维度在flatten range 的 中间，预期结果切分维度被推导成复制

**Review by JZ-LIANG — DISMISSED** [review]
LGTM

File: test/cpp/auto_parallel/spmd_rule_test.cc (hunk 1 above)
Comment (jeff41404) [reply test/cpp/auto_parallel/spmd_rule_test.cc:1884]: ok, the unittest case of shard dimension in the middle range of flatten has been added.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**